### DISCUSSION TITLE
Make Rejected Translations More Accessible

### DISCUSSION THREAD

**Discussion by Kacarott** [codewars/codewars.com#2796] [Ideas]
Currently, there are a *huge* number of pending translations which have various issues preventing their approval. This large number makes it much more difficult to identify older translations which are actually in an acceptable state.

A simple solution is that old translations with issues should be rejected. However there is a tendency against rejecting translations due to fear of losing work. "This translation has an old version and lacks random tests, but what if someone was planning on repairing it?". Of course, a rejected translation is not *deleted*, however it becomes much less accessible. Without a known link, the only way to find any rejected translations is in the translation panel of a kata, mixed in with all the approved translations.

My suggestion is that by making rejected translations more discoverable, it will be far more acceptable to simple reject faulty translations, making content management much easier. In order to make rejected translations more accessible, I have two ideas:
- Separate rejected translations from approved translations in the translation panel of a kata.
- In the "Translation editor", add a panel containing previously rejected translations in that language, allowing translators to easily reference them while authoring a new translation.

**Comment by hobovsky** [comment]
I think your question is an example of an XY problem and your question is not really "How to make rejected translations more accessible?", but it rather is "How to reduce a backlog of pending translations?" That's why I share some thoughts on a completely different question than you asked :)

**WARNING:** the post below might be considered off-topic :)

I can see more than one problem here, and I think a good question to answer first would be: why are translations even laying there, unapproved? There would be no problem of old translations, if there were no old translations :) What makes pending translations grow old? I have a couple of ideas:

- System allows translations to grow old. If translations were auto-rejected after a year and/or after collecting a couple of downvotes, the hoarder's mentality would not even get a chance to kick in.
- Pending translations are difficult to discover - I think this is 50% true, because after solving a kata users see pending translations, but I agree that this might be not enough. There are announcement posts, which help, and there's the "Pending translations" view, but it is (intentionally) accessible only to selection of users. Additional factor can be the fact that users need to solve a kata to view translations, and translation on difficult kata can go unnoticed for long.
- Users are not interested in reviewing and approving translations in general - nah don't think so, there are users very dedicated to reviewing, who do the task really well.
- Users want to review translations, but the rules are too difficult to follow - I've heard this complaint a couple of times, but I am not sure I agree with it. There happen to be (overly?) enthusiastic users who click the green button a couple of times, then the fact that the approved translations were not good (for one reason or another) is pointed out, and they get bitter over the fact that they wanted to do good, but others say they did bad. I _would_ consider this as one of the reasons for translations getting stale, but my experience says otherwise: there is many, many good and dedicated translators and reviewers who got familiar with the guidelines, know how to apply them to reviewed translations, and follow them, creating and approving heaps of translations on a regular basis. My feeling is that quality of incoming and approved translations increased a really, really lot and I am really impressed by efforts many translators go through to make their work good. I consider the _"rules are too strict"_ not a really good argument, mainly used by people who want to avoid effort.
- Translations are difficult to approve for technical reasons - this might play a large role here. Translations are not very easy to verify, and not easy to merge due to merge conflicts. I'd hope that possibility of training on translations would significantly help. Contributor bug is also a PITA, I hope it will get fixed.
- Approvers do not want to bear a responsibility of approval - I personally think this is a big problem, because I heard a couple of times already that someone does not want to click the green button to not do something wrong. I am not sure what would be the best approach here, but I think that it would be good just if approvers followed the guidelines and listened to feedback of others. Also a possibility of reducing the responsibility in some way, for example voting on a translation as approvable would be nice: "Is this translation good to be approved? Please click 👎 or 👍 ". It's easier to approve something when you see that 5 or 10 other people consider it good, or reject it when others consider it bad.
- Quality of the translated kata makes approval difficult - I think this is very important point, and for me it is _the_ reason why I do not approve translations: some kata are such a mess that it's difficult to decide whether a pending translation even addresses the same problem. Inconsistent performance, requirements, non-translatable idioms (mixed types, yay!), bad practices, all this makes me just sit and wonder, "what the hell is this kata exactly about, and does the translation even address the task?". I am also not willing to approve a translation to a kata which relies on wrong practices (_"round the answer to X decimal places"_) just to not make the practice spread.
- Reviewers do not want to reject bad translations for various reasons (yay, we finally got to this point) - this is a valid concern, but I think it's an answer to a very late problem which can be prevented by other means and not even appear. There can be also different reasons for sticking to bad translations, like users who do not want to give negative feedback to translators to not hurt their feelings. A couple of times I've seen a translator who got butthurt over their translation being rejected, commented with issue reports, or forked with a bugfix. Making a translation hidden is also an issue, but as I tried to prove, it's just a part of the general problem. I just think that making rejected translations more discoverable will not prevent the bad, rejection worthy translations from coming over and over again. It will also not stop the backlog of good translations from growing.

**Comment by hobovsky** [comment]
> * Separate rejected translations from approved translations in the translation panel of a kata.

I like this idea, and I think that "Translations" tab could present more details:
- Status of closed translations is not shown (approved/rejected/deprecated)
- It could be nice to see if a translation is a bug fix (for example, a fork of a kata), or a fork of a fork, or new translation. Maybe forks of forks could be displayed in a hierarchical way?
- Specifically about rejected translations, I think they are not important if another translation to the same language is already approved. It might be good to consider only rejected translations to unavailable languages as relevant.

> * In the "Translation editor", add a panel containing previously rejected translations in that language, allowing translators to easily reference them while authoring a new translation.

Would it fit on screen? Besides, a rejected translations could still be forked to be used as a starting point for a new one?
